src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -119,37 +117,31 @@
     
     private String originalHostNameInCertificate = null;
 
-    static class Sha1HashKey {
-        private byte[] bytes;
+    static class CityHash128Key {
+        private long[] segments;
+        private int hashCode;
 
-        Sha1HashKey(String sql,
+        CityHash128Key(String sql,
                 String parametersDefinition) {
             this(String.format("%s%s", sql, parametersDefinition));
         }
 
-        Sha1HashKey(String s) {
-            bytes = getSha1Digest().digest(s.getBytes());
+        CityHash128Key(String s) {
+            segments = CityHash.cityHash128(s.getBytes(), 0, s.length());
```

Comment:
Performance of this can be improved, especially in the case of large SQL strings.  ``s.getBytes()`` ends up calling ``StringCoding.encode(...)``, which converts the internal String chars to bytes using the platform default encoding.

I recommend using ``s.getBytes(int srcBegin, int srcEnd, byte dst[], int dstBegin)``, which, while deprecated (but will never be removed), has *substantially* higher performance.
```java
CityHash128Key(String s) {
   byte[] bytes = byte[s.length()];
   s.getBytes(0, s.length(), bytes, 0);
   segments = CityHash.cityHash128(bytes, 0, s.length());
}
```

Reply:
s.getBytes(int srcBegin, int srcEnd, byte dst[], int dstBegin) truncates the upper 8 bits of Java's internal Unicode-16 representation. Since this is SQL before inserting parameter values, it should be mostly 7-bit ASCII -- the only exception would be database names/table names/column names, at least some of which can be full Unicode. For European languages, their Unicode-16 spaces tend to be 256 characters wide and the odds of this causing a spurious collision seem very small -- but I'd be less confident if the language in question was, say, Chinese, that you couldn't have two SQLs that differed only by one ideogram in a table/column name, and the two ideograms in question happened to have the same lower 8 bits. It's not very likely, but it's significantly less astronomically unlikely than a 128-bit hash collision.

Reply:
So yes, we need to avoid character reencoding, but I think we also need to avoid throwing away hash entropy while we do so. I was wondering about something like getBytes(Characterset.UTF_16BE) -- if it matches the internal representation used by String, will it skip reencoding?

Reply:
@RDearnaley It appears only option would be a custom ``Charset`` and ``CharsetEncoder``.  It is *possible*, as there is a ``CharsetProvider`` SPI available.  The "raw encoder" would simply "encode" chars as pairs of bytes (upper 8-bits, lower 8-bits).  The cost is a 2x sized byte array due to the naive approach, but possibly worth it as the byte array is quickly garbage collected, or could potentially be cached as a ``WeakReference`` in a ``ThreadLocal`` inside of the ``CityHash128Key`` class.

Reply:
How strange -- yes, taking a look in the Java code for them, it appears that none of UTF_16BE, UTF_16 etc make use of the obvious optimization that one of them has to be a no-op. I'm wondering if someone might have already created a raw double byte encoding to speed this up -- it seems a fairly obvious performance optimization for cases like this where you care about the entropy content rather than the specific values. If not, it might make a good addition to the Java version of CityHash

Reply:
There is another possible approach whose speed would be worth testing:

    private static byte[] toBytes( final String str )
    {
        final byte[] buf= new byte[ 2 * str.length() ];
        for (int i = 0, j = 0; i < str.length(); i++) {
            char c = str.charAt( i );
            buf[j++] = (byte) c;
            buf[j++] = (byte) (c >>> 8);  // Or >> might be faster?
        }
        return buf;
    }

or just inline this since we only use it once.

Reply:
Of course it would work, but getChars() is going to create an additional copy, incurring a CPU hit as well as increasing GC pressure.  Regarding inlining, probably unnecessary as the JIT will inline it when it gets hot.

I think a custom encoder would offer better performance, as the internal char array is passed without copy.

Reply:
Would the str.charAt(i) also be making a copy, or would that just go on the stack/in a register?

Reply:
str.charAt(i) is just a simple array access, with the value passed on the stack.  If the compiler inlines the call, it could be a viable option.  You'd have to find it in the output of ``-XX:+UnlockDiagnosticVMOptions -XX:+LogCompilation -XX:+PrintCompilation``.